### PR TITLE
GOOW-3 updated tile generation

### PR DESCRIPTION
**Pull request by beserkzombie**

### CODE REVIEW

File: TileMap.gd [hunk 1]
```diff
@@ -5,34 +5,50 @@ var overworld
...
+	BRCorner,	# 0
+	Door,	# 1
+	Floor,	# 2
+	VWall,	# 3
+	HWall,	# 4
+	TLCorner,	# 5
+	TRCorner, #6
+	TWall, #7
+	BLCorner, #8
+	Ladder
```
Comment (demwilson) [inline TileMap.gd]: I'd suggest using full words for describing enums.
Example:
`BRCorner`
would be:
`Bottom_Right_Corner`

File: TileMap.gd [hunk 2]
```diff
@@ -66,265 +82,614 @@ func _on_Player_collided(collision):
...
+		add_room(currentSpot, needsCorridor)
 
-	# Place end ladder
-	var end_room = rooms.back()
-	var ladder_x = end_room.position.x + 1 + randi() % int(end_room.size.x - 2)
-	var ladder_y = end_room.position.y + 1 + randi() % int(end_room.size.y - 2)
-	set_tile(ladder_x, ladder_y, Tile.Ladder)
+		if needsCorridor:
+			#close blocked off sections
+			#TODO: number of exits is based on corridor scene will need to get that value somewhere
```
Comment (demwilson) [inline TileMap.gd]: If this TODO is an issue, you should create an issue for it. Otherwise we should delete this comment.

File: TileMap.gd [hunk 3]
```diff
@@ -66,265 +82,614 @@ func _on_Player_collided(collision):
...
-	var points = graph.get_points()
-	var start = points.pop_back()
-	for point in points:
-		var path = graph.get_point_path(start, point)
-		if !path:
-			return false
+func place_player_start(startingSpot):
+	var start_room = startingSpot
+	var player_x = start_room.x + 4 - randi() % 2
+	var player_y = start_room.y + 4 - randi() % 2
```
Comment (demwilson) [inline TileMap.gd]: What do the magic numbers `4` and `2` represent?

File: TileMap.gd [hunk 4]
```diff
@@ -66,265 +82,614 @@ func _on_Player_collided(collision):
...
+func place_exit(startingSpot):
+	var end_room
+	var cellIndex
+	var ladderPlaced = false	
+	
+	while !ladderPlaced: 
+		if startingSpot == MapCornerPoints.TopLeft:
+			#set ladder bottomRight
+			end_room = MapCornerPoints.BottomRight
+			cellIndex = self.get_cell(end_room.x + 4, end_room.y + 4)
```
Comment (demwilson) [inline TileMap.gd]: What does the magic number `4` represent?

File: TileMap.gd [hunk 5]
```diff
@@ -66,265 +82,614 @@ func _on_Player_collided(collision):
...
+				ladderPlaced = true
+		else:
+			#set ladder topRight
+			end_room = MapCornerPoints.TopRight
+			cellIndex = self.get_cell(end_room.x + 4, end_room.y + 4)
+			if cellIndex != -1:
+				ladderPlaced = true
+			
+	var ladder_x = end_room.x + 4 - randi() % 2
+	var ladder_y = end_room.y + 4 - randi() % 2
```
Comment (demwilson) [inline TileMap.gd]: What do the magic numbers `4` and `2` represent?

File: TileMap.gd [hunk 6]
```diff
@@ -66,265 +82,614 @@ func _on_Player_collided(collision):
...
+	
+	#map the scene to the tilemap
+	if isCorridor:
+		scene = Corridor_1111.instance()
+		roomsTileMap = scene.get_node("TileMap")
+		usedCells = roomsTileMap.get_used_cells()
+		for usedCell in usedCells:
+			var cellindex = roomsTileMap.get_cell(usedCell.x, usedCell.y)
+			if cellindex == Tile.BRCorner:
+				if usedCell.x == 2 && usedCell.y == 6:
```
Comment (demwilson) [inline TileMap.gd:227]: What do the magic numbers `2` and `6` represent?

File: TileMap.gd [hunk 7]
```diff
@@ -66,265 +82,614 @@ func _on_Player_collided(collision):
...
+					#rotate tile 270 degree to right
+					set_tile(topLefPosition.x + usedCell.x, topLefPosition.y + usedCell.y, cellindex, false, true, true)
+				elif usedCell.x == 6 && usedCell.y == 6:
```
Comment (demwilson) [inline TileMap.gd:230]: What do the magic numbers `6` represent?

**Review by demwilson — CHANGES_REQUESTED** [review]
We already talked about abstracting some of the repetitive code to its own function, but I think just creating an issue for it is good enough for this PR.

File: TileMap.gd (hunk 1 above)
Comment (beserkzombie) [reply TileMap.gd]: you mean BottomRightCorner?

**Review by demwilson — APPROVED** [review]

**Pull request merged**